### PR TITLE
KOGITO-4907:PoC around invoking runtimes from Data Index using rest-client

### PR DESCRIPTION
**Pull request by nmirasch**
Many thanks for submitting your Pull Request :heart:! 

Jira: https://issues.redhat.com/browse/KOGITO-4907
<summary>
How to retest this PR or trigger a specific build:
</summary>

* <b>Pull Request</b>  
  Please add comment: <b>Jenkins retest this</b>
 
* <b>Quarkus LTS checks</b>  
  Please add comment: <b>Jenkins run LTS</b>

* <b>Native checks</b>  
  Please add comment: <b>Jenkins run native</b>

</details>

### CODE REVIEW

File: data-index/data-index-service/data-index-service-common/src/main/java/org/kie/kogito/index/api/DataIndexServiceClient.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ * limitations under the License.
+ */
+package org.kie.kogito.index.api;
+
+import java.io.Closeable;
+import java.util.List;
+
+import org.kie.kogito.index.model.ProcessInstance;
+
+public interface DataIndexServiceClient extends Closeable {
```
Comment (wmedvede) [inline data-index/data-index-service/data-index-service-common/src/main/java/org/kie/kogito/index/api/DataIndexServiceClient.java]: Adding this code in the data-index-service-common module will make it hard/impossible to use it in other places or interested parties. Note that this module contains the data index service core, which mean that any application trying to use this client api, that adds this module as maven dependency, will automatically incorporate the data index core and for example the data index endpoints, etc.

File: data-index/data-index-service/data-index-service-common/src/main/java/org/kie/kogito/index/api/DataIndexServiceClientConfig.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.kie.kogito.index.api;
+
+import java.net.URL;
+
+import org.kie.kogito.index.api.config.ServiceClientConfig;
+
+public class DataIndexServiceClientConfig extends ServiceClientConfig {
```
Comment (wmedvede) [inline data-index/data-index-service/data-index-service-common/src/main/java/org/kie/kogito/index/api/DataIndexServiceClientConfig.java]: Same as above for this class

File: data-index/data-index-service/data-index-service-common/src/main/java/org/kie/kogito/index/api/DataIndexServiceClientFactory.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.kie.kogito.index.api;
+
+public interface DataIndexServiceClientFactory {
```
Comment (wmedvede) [inline data-index/data-index-service/data-index-service-common/src/main/java/org/kie/kogito/index/api/DataIndexServiceClientFactory.java]: Same here

File: data-index/data-index-service/data-index-service-common/src/main/java/org/kie/kogito/index/api/GraphQLServiceClient.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.kie.kogito.index.api;
+
+import java.io.Closeable;
+
+public interface GraphQLServiceClient extends Closeable {
```
Comment (wmedvede) [inline data-index/data-index-service/data-index-service-common/src/main/java/org/kie/kogito/index/api/GraphQLServiceClient.java]: Same here

File: data-index/data-index-service/data-index-service-common/src/main/java/org/kie/kogito/index/api/mp/GraphQLServiceClientRest.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+
+import com.fasterxml.jackson.databind.node.ObjectNode;
+
+import static javax.ws.rs.core.MediaType.APPLICATION_JSON;
+
+public interface GraphQLServiceClientRest extends Closeable {
+
+    @Path("/graphql")
```
Comment (wmedvede) [inline data-index/data-index-service/data-index-service-common/src/main/java/org/kie/kogito/index/api/mp/GraphQLServiceClientRest.java]: looks like clients won't be able to change the graphql server path

**Review by wmedvede — COMMENTED** [review]
Hi  @nmirasch , I left some comments that I believe makes sense to consider.
Have look please.

Thanks

**Comment by nmirasch** [comment]
@wmedvede yes the idea is that all included in the api package will be outside. We need to decide where is the best place. Right now I'm focused in the integration test to be able to try the added data-index-service mutations and the service interactions.

@cristianonicolai I've added the integration test even they are failing. Could you have a look? Thanks!

File: apps-integration-tests/integration-tests-data-index-service/integration-tests-data-index-service-common/src/main/java/org/kie/kogito/index/resources/KogitoServiceRandomPortTestResource.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+    @Override
+    public String getResourceName() {
+        return NAME;
+    }
+
+    @Override
+    public void start() {
+        httpPort = SocketUtils.findAvailablePort();
+        Testcontainers.exposeHostPorts(httpPort);
```
Comment (nmirasch) [inline apps-integration-tests/integration-tests-data-index-service/integration-tests-data-index-service-common/src/main/java/org/kie/kogito/index/resources/KogitoServiceRandomPortTestResource.java:38]: @cristianonicolai  without this 'Testcontainers.exposeHostPorts(httpPort);' kafka is able to start but the kogito.service.url can't be resolved.

Comment (cristianonicolai) [reply apps-integration-tests/integration-tests-data-index-service/integration-tests-data-index-service-common/src/main/java/org/kie/kogito/index/resources/KogitoServiceRandomPortTestResource.java:38]: @nmirasch seems like a known issue with testcontainers, maybe try the suggested workaround: https://github.com/testcontainers/testcontainers-java/issues/3187#issuecomment-827104113

Comment (nmirasch) [reply apps-integration-tests/integration-tests-data-index-service/integration-tests-data-index-service-common/src/main/java/org/kie/kogito/index/resources/KogitoServiceRandomPortTestResource.java:38]: I'll try, thanks!

**Comment by nmirasch** [comment]
@cristianonicolai I've modified the mutation code to not to start a new thread. I've been added some tests and I've created a new data-index-api to contain all the api part, just to add some test  on the api avoiding add dependences to data-index-service-common just to try the connections... 
As we talked this package needs to be placed outside ... Please, when you have any chance, take a look. Any thought is very welcome, Thanks!

**Comment by nmirasch** [comment]
Jenkins please retest this

**Comment by nmirasch** [comment]
Jenkins please retest this

**Comment by nmirasch** [comment]
Jenkins retest this

**Comment by nmirasch** [comment]
Jenkins retest this

**Comment by nmirasch** [comment]
Closing this regarding the dev is included here and this POC have been used to generate this ..https://github.com/kiegroup/kogito-apps/pull/929

**Pull request closed without merging**